File: Classes/ElapsedTimeViewController.m
//
//  ElapsedTimeViewController.m
//  KidsTimeFun
//
//  Created by Jagmeet Chawla on 4/12/09.
//  Revised by Krishna Narayan on 9/17/09
//		--Updated graphics
//		--Added tell a friend
//		--Changed app name to Kids Learn To Tell Time
//  Copyright 2009-2013 NSC Partners LLC. Copyright 2014 One Step Ahead Apps, LLC. All rights reserved.
//

#import "ElapsedTimeViewController.h"
#import "ClockView.h"
#import "RandomInteger.h"
#import "KidsTimeFunAppState.h"
#import "AudioPlayer.h"

@implementation ElapsedTimeViewController

@synthesize delegate;
@synthesize isRight;
@synthesize activity;
@synthesize activityType;
@synthesize activityLevel;
@synthesize timeOffset;
@synthesize answerIndex;

 // The designated initializer.  Override if you create the controller programmatically and want to perform customization that is not appropriate for viewDidLoad.
- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
	[super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

/*
// Implement loadView to create a view hierarchy programmatically, without using a nib.
- (void)loadView {
}
*/


// Implement viewDidLoad to do additional setup after loading the view, typically from a nib.
- (void)viewDidLoad {
	//Build first clock
	clockView = [[ClockView alloc] initWithFrame:clockContainerView.bounds];
	clockView.backgroundColor = [UIColor colorWithRed:0.0f green:0.0f blue:0.0f	alpha:0.0f];
	
	//Build 2nd clock
	if (self.activity == kActElapsedTime) {
		clockView2 = [[ClockView alloc] initWithFrame:clockContainerView.bounds];
		clockView2.backgroundColor = [UIColor colorWithRed:0.0f green:0.0f blue:0.0f alpha:0.0f];
	}
	
	//initialize intervals and range based on difficulty settings
	int timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices;
	switch (self.activityLevel) {
		case kActLevelYellowBelt:
			numberOfChoices = 2;
			timeInterval = 30;
			timeRangeLow = 30;
			timeRangeHigh = 120;
			break;
		case kActLevelGreenBelt:
			numberOfChoices = 3;
			timeInterval = 15;
			timeRangeLow = 15;
			timeRangeHigh = 240;
			break;
		case kActLevelRedBelt:
			numberOfChoices = 4;
			timeInterval = 5;
			timeRangeLow = 5;
			timeRangeHigh = 360;
			break;
		case kActLevelBlackBelt:
			numberOfChoices = 4;
			timeInterval = 1;
			timeRangeLow = 5;
			timeRangeHigh = 719;
			break;
		default:
			break;
	}
	
	//set number of choices on UI
	switch (numberOfChoices) {
		case 2:
			choices4.hidden = YES;
			choices3.hidden = YES;
			choices2.hidden = NO;
			choices = choices2;
			break;
		case 3:
			choices4.hidden = YES;
			choices3.hidden = NO;
			choices2.hidden = YES;
			choices = choices3;
			break;
		case 4:
			choices4.hidden = NO;
			choices3.hidden = YES;
			choices2.hidden = YES;
			choices = choices4;
			break;
		default:
			break;
	}
	
	//Load first/main clock with Random values
	//Hours first
	randomNumber = [[RandomInteger alloc] initWithRange:1 To:12];
	clockView.hours = randomNumber.randomInteger;
	//Then AM/PM
	clockView.PM = ([randomNumber nextRandomIntegerInRange:0 To:1])?YES:NO;
	//rounded to nearest multiple of timeInterval
	clockView.minutes = ([randomNumber nextRandomIntegerInRange:0 To:59] / timeInterval)*timeInterval;
	//No seconds for now
	clockView.seconds = 0;
	clockView.showSeconds = NO;
	//other clock properties
	clockView.showClockAsAnalog = YES;
	clockView.showMinutesOffsetInHoursHand = YES;
	clockView.showAMPM = NO;
	clockView.showDayNight = NO;
	//add clock as a view in the container
	[clockContainerView addSubview:clockView];
	//NSLog(@"First: %f %f %f %f",clockView.frame.origin.x,clockView.frame.origin.y,clockView.frame.size.width,clockView.frame.size.height);
		
	//Then ask the question
	int randomTimeOffset, randomHoursOffset, randomMinutesOffset;
	 labelQuestion.text = @"How much time has passed?";
	[[AudioPlayer getInstance] playAudioFile:@"how_much_time_has_past"];
	
	//create timeoffset for clock 2
	randomTimeOffset = [randomNumber nextRandomIntegerInRange:timeRangeLow To:timeRangeHigh];
	randomTimeOffset = (randomTimeOffset/timeInterval)*timeInterval;
	randomHoursOffset = randomTimeOffset/60;
	randomMinutesOffset = randomTimeOffset-(randomHoursOffset*60);
	
	//this offset is the answer
	int answerHours, answerMinutes;
	answerHours = randomHoursOffset;
	answerMinutes = randomMinutesOffset;
	
	//hours and minutes for 2nd clock
	int clock2hours, clock2minutes;
	clock2minutes = (int)clockView.hours*60+(int)clockView.minutes;
	clock2minutes = clock2minutes +randomTimeOffset;
	if (clock2minutes < 0) clock2minutes = clock2minutes+720;
	if (clock2minutes > 720) clock2minutes = clock2minutes-720;
	clock2hours = clock2minutes/60;
	clock2minutes = clock2minutes-(clock2hours*60);
	if (clock2hours == 0) clock2hours = 12;
	
	//Load 2nd clock
	//Hours first
	clockView2.hours = clock2hours;
	//Then AM/PM
	clockView2.PM = clockView.PM;
	//rounded to nearest multiple of timeInterval
	clockView2.minutes = clock2minutes;
	//No seconds for now
	clockView2.seconds = 0;
	clockView2.showSeconds = NO;
	//other clock properties
	clockView2.showClockAsAnalog = YES;
	clockView2.showMinutesOffsetInHoursHand = YES;
	clockView2.showAMPM = NO;
	clockView2.showDayNight = NO;
	//add clock as a view in the container
	[clockContainerView2 addSubview:clockView2];
	
	//load random answer choices
	int h=0, m=0, prev_h=0, prev_m=0;
	NSString* answerTitle =  (answerHours == 0) ? [NSString stringWithFormat:@"%02i min", answerMinutes] : [NSString stringWithFormat:@"%i:%02i", answerHours, answerMinutes];

	for (int i = 0; i<choices.numberOfSegments; i++) {
		BOOL newNumberGenerated = NO;
		NSString* choiceTitle;
		while (!newNumberGenerated) {
			//generate random time
			h = [randomNumber nextRandomIntegerInRange:((timeRangeLow/60)==0)?1:timeRangeLow To:(timeRangeHigh/60)];
			m = ([randomNumber nextRandomIntegerInRange:0 To:59]/timeInterval)*timeInterval;
			//if same as previous or answer, generate new random minute to change it
			int loopLimit=0;
			while ((((prev_h*60+prev_m) == (h*60+m))||((h*60+m) == (answerHours*60+answerMinutes)))&&(loopLimit<3)) {
				m = (randomNumber.randomInteger/timeInterval)*timeInterval;
				++loopLimit;
			}
			prev_m = m;
			prev_h = h;
			choiceTitle = (h == 0) ? [NSString stringWithFormat:@"%02i min", m] : [NSString stringWithFormat:@"%i:%02i", h, m];
			if (i == 0) {
				newNumberGenerated = YES;
			}
			else 
			{
				for (int j = 0; j < i; j++) 
				{
					if ([[choices titleForSegmentAtIndex:j] isEqualToString:choiceTitle] || [answerTitle isEqualToString:choiceTitle]) {
						newNumberGenerated = NO;
						break;
					}
					else {
						newNumberGenerated = YES;
					}

				}
			}
		}
		[choices setTitle:choiceTitle forSegmentAtIndex:i];
	}
	//load answer in a random index, but remember where it is
	randomNumber.rangeLow = 0;
	randomNumber.rangeHigh =(int) choices.numberOfSegments-1;
	self.answerIndex = randomNumber.randomInteger;
//	if (answerHours == 0) [choices setTitle:[NSString stringWithFormat:@"%02i min", answerMinutes] forSegmentAtIndex:self.answerIndex];
//		else [choices setTitle:[NSString stringWithFormat:@"%i:%02i", answerHours, answerMinutes] forSegmentAtIndex:self.answerIndex];
	[choices setTitle:answerTitle forSegmentAtIndex:self.answerIndex];
	//now load super
    [super viewDidLoad];
}

/*
// Override to allow orientations other than the default portrait orientation.
- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation {
    // Return YES for supported orientations
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}
*/

- (void)didReceiveMemoryWarning {
	// Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
	
	// Release any cached data, images, etc that aren't in use.
}

- (void)viewDidUnload {
	// Release any retained subviews of the main view.
	// e.g. self.myOutlet = nil;
}


- (void)dealloc {
	[randomNumber release];
	[clockContainerView release];
	[clockContainerView2 release];
	[clockView2 release];
	[clockView release];
	if (self.delegate != self) delegate = nil;
	[choices release];
	[rightOrWrong release];
    [super dealloc];
}

#pragma mark -
#pragma mark Custom Methods

- (IBAction) choicesValueChanged {
	choices.enabled = NO;
	choices.hidden = YES;
	if (choices.selectedSegmentIndex == self.answerIndex) {
		isRight = YES;
		rightOrWrong.image = [UIImage imageNamed:@"GoodJob.png"];
		rightOrWrong2.image = [UIImage imageNamed:@"Right.png"];
		rightOrWrong2.hidden = NO;
		[[AudioPlayer getInstance] playCorrectWrong:YES];
		[NSTimer scheduledTimerWithTimeInterval:1.5 target:self selector:@selector(rightAnswer) userInfo:NULL repeats:NO];
	}
	else {
		isRight = NO;
		++wrongCounter;
		rightOrWrong2.image = [UIImage imageNamed:@"Wrong.png"];
		rightOrWrong.image = [UIImage imageNamed:@"TryAgain.png"];
		[[AudioPlayer getInstance] playCorrectWrong:NO];
		if (self.activityType == kActTypeTimed) {
			rightOrWrong2.hidden = NO;
			rightOrWrong.hidden = YES;
		}
		else {
			rightOrWrong2.hidden = YES;
			rightOrWrong.hidden = NO;
		}
		NSLog(@"Wrong Answer");
		[NSTimer scheduledTimerWithTimeInterval:1.5 target:self selector:@selector(wrongAnswer) userInfo:NULL repeats:NO];
	}
}

- (void) rightAnswer {
	if (wrongCounter > 0) {
		isRight = NO;
		wrongCounter = 0;
	}
	if ([delegate conformsToProtocol:@protocol(DismissActivityDelegate)]) [delegate didDismissActivity:self];
}

- (void) wrongAnswer {
	if (self.activityType != kActTypeTimed) {
		rightOrWrong.image = nil;
		rightOrWrong2.image = nil;
		rightOrWrong2.hidden = YES;
		choices.enabled = YES;
		choices.hidden = NO;
		[self.view setNeedsDisplay];
	}
	else {
		if ([delegate conformsToProtocol:@protocol(DismissActivityDelegate)]) [delegate didDismissActivity:self];
	}

}

@end

File: Classes/ElapsedTimeViewController.swift
import UIKit

@objc(ElapsedTimeViewController)
class ElapsedTimeViewController: BaseViewController {

    weak var delegate: DismissActivityDelegate?
    private(set) var isRight = false
    var activity: Int32 = 0
    var activityType: Int32 = 0
    var activityLevel: Int32 = 0
    var timeOffset: Int32 = 0
    var answerIndex: Int = 0

    @IBOutlet private weak var clockContainerView: UIView!
    @IBOutlet private weak var clockContainerView2: UIView!
    @IBOutlet private weak var choices2: UISegmentedControl!
    @IBOutlet private weak var choices3: UISegmentedControl!
    @IBOutlet private weak var choices4: UISegmentedControl!
    @IBOutlet private weak var rightOrWrong: UIImageView!
    @IBOutlet private weak var rightOrWrong2: UIImageView?
    @IBOutlet private weak var labelQuestion: UILabel!

    private var clockView: ClockView!
    private var clockView2: ClockView!
    private var randomNumber: RandomInteger!
    private var choices: UISegmentedControl!
    private var wrongCounter = 0

    override func viewDidLoad() {
        clockView = ClockView(frame: clockContainerView.bounds)
        clockView.backgroundColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0)

        if activity == kActElapsedTime {
            clockView2 = ClockView(frame: clockContainerView.bounds)
            clockView2.backgroundColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0)
        }

        let timeInterval: Int, timeRangeLow: Int, timeRangeHigh: Int, numberOfChoices: Int
        switch activityLevel {
        case kActLevelYellowBelt: (timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices) = (30, 30, 120, 2)
        case kActLevelGreenBelt:  (timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices) = (15, 15, 240, 3)
        case kActLevelRedBelt:    (timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices) = (5, 5, 360, 4)
        case kActLevelBlackBelt:  (timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices) = (1, 5, 719, 4)
        default:                  (timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices) = (30, 30, 120, 2)
        }

        switch numberOfChoices {
        case 2: choices4.isHidden = true;  choices3.isHidden = true;  choices2.isHidden = false; choices = choices2
        case 3: choices4.isHidden = true;  choices3.isHidden = false; choices2.isHidden = true;  choices = choices3
        default:choices4.isHidden = false; choices3.isHidden = true;  choices2.isHidden = true;  choices = choices4
        }

        randomNumber = RandomInteger(range: 1, to: 12)
        clockView.hours = Float(randomNumber.randomInteger)
        clockView.PM = randomNumber.nextRandomInteger(inRange: 0, to: 1) != 0
        clockView.minutes = Float((randomNumber.nextRandomInteger(inRange: 0, to: 59) / timeInterval) * timeInterval)
        clockView.seconds = 0; clockView.showSeconds = false
        clockView.showClockAsAnalog = true; clockView.showMinutesOffsetInHoursHand = true
        clockView.showAMPM = false; clockView.showDayNight = false
        clockContainerView.addSubview(clockView)

        labelQuestion.text = "How much time has passed?"
        AudioPlayer.getInstance().playAudioFile("how_much_time_has_past")

        let randomOffset = (randomNumber.nextRandomInteger(inRange: timeRangeLow, to: timeRangeHigh) / timeInterval) * timeInterval
        let answerHours = randomOffset / 60
        let answerMinutes = randomOffset - answerHours * 60

        var c2m = Int(clockView.hours) * 60 + Int(clockView.minutes) + randomOffset
        if c2m < 0 { c2m += 720 }; if c2m > 720 { c2m -= 720 }
        var c2h = c2m / 60; let c2mins = c2m - c2h * 60
        if c2h == 0 { c2h = 12 }
        clockView2.hours = Float(c2h); clockView2.PM = clockView.PM
        clockView2.minutes = Float(c2mins); clockView2.seconds = 0; clockView2.showSeconds = false
        clockView2.showClockAsAnalog = true; clockView2.showMinutesOffsetInHoursHand = true
        clockView2.showAMPM = false; clockView2.showDayNight = false
        clockContainerView2.addSubview(clockView2)

        let answerTitle = answerHours == 0 ? String(format: "%02i min", answerMinutes) : "\(answerHours):\(String(format: "%02i", answerMinutes))"
        var prevH = 0, prevM = 0
        for i in 0..<choices.numberOfSegments {
            var h = 0, m = 0
            repeat {
                h = randomNumber.nextRandomInteger(inRange: max(timeRangeLow / 60, 1), to: timeRangeHigh / 60)
                m = (randomNumber.nextRandomInteger(inRange: 0, to: 59) / timeInterval) * timeInterval
            } while (h * 60 + m == prevH * 60 + prevM || h * 60 + m == answerHours * 60 + answerMinutes)
            prevH = h; prevM = m
            let t = h == 0 ? String(format: "%02i min", m) : "\(h):\(String(format: "%02i", m))"
            choices.setTitle(t, forSegmentAt: i)
        }

        randomNumber.rangeLow = 0; randomNumber.rangeHigh = choices.numberOfSegments - 1
        answerIndex = randomNumber.randomInteger
        choices.setTitle(answerTitle, forSegmentAt: answerIndex)

        super.viewDidLoad()
        edgesForExtendedLayout = []
    }

    @IBAction func choicesValueChanged() {
        choices.isEnabled = false; choices.isHidden = true
        if choices.selectedSegmentIndex == answerIndex {
            isRight = true
            rightOrWrong.image = UIImage(named: "GoodJob")
            rightOrWrong2?.image = UIImage(named: "Right"); rightOrWrong2?.isHidden = false
            AudioPlayer.getInstance().playCorrectWrong(true)
            Timer.scheduledTimer(timeInterval: 1.5, target: self, selector: #selector(rightAnswer), userInfo: nil, repeats: false)
        } else {
            isRight = false; wrongCounter += 1
            rightOrWrong2?.image = UIImage(named: "Wrong"); rightOrWrong.image = UIImage(named: "TryAgain")
            AudioPlayer.getInstance().playCorrectWrong(false)
            rightOrWrong2?.isHidden = true; rightOrWrong.isHidden = false
            Timer.scheduledTimer(timeInterval: 1.5, target: self, selector: #selector(wrongAnswer), userInfo: nil, repeats: false)
        }
    }

    @objc private func rightAnswer() {
        if wrongCounter > 0 { isRight = false; wrongCounter = 0 }
        delegate?.didDismissActivity(self)
    }

    @objc private func wrongAnswer() {
        if activityType != kActTypeTimed {
            rightOrWrong.image = nil; rightOrWrong2?.image = nil
            rightOrWrong2?.isHidden = true; choices.isEnabled = true; choices.isHidden = false
            view.setNeedsDisplay()
        } else {
            delegate?.didDismissActivity(self)
        }
    }
}
